### ISSUE TITLE
Startup hangs and CPU lockup

### ISSUE THREAD

**Issue by mstormi** [openhab/openhab-distro#531]
Hi all,
I keep encountering the problem that OH fails to startup.
At some point in the startup process, Java stays at 100%, but there is no more progress.
My strong guess is that to load/instantiate specific bindings takes too long and we run into a karaf timeout. Other users are probably not seeing this because they have a faster machine, less rules, or both.

I've encountered at least two different versions of the problem (maybe it's 2 different ones):
1) Most common is to see java stay at 100+ % CPU all day.
This is on a 4-core Pi2, so a single CPU core is hogged. CPU goes up as high as 350% at times during startup (which is ok).
Most things seem to work (not sure though if ALL of them, hard to see), but as the main java thread keeps running at 100%, everything is way too slow to be usable.
2) Java is well below 100%, but OH fails to even load OH binding bundles as defined in addons.cfg.
The "Eclipse SmartHome Configuration Discovery" bundle remains in "Waiting" state, sometimes (not sure if always) the "Eclipse SmartHome Bonjour/MDS Service Discovery Bundle" as well.

Main problem is: I don't know how to methodically approach this for debugging.
What log settings should I apply? I just enabled org.apache.karaf for now.

But the key question is: how can I find out what entity is eating CPU ?

On a sidenote: whenever a bundle is in Waiting state, how do I find out what it's waiting for ?

My various trial-and-error attempts to pin down the 'root cause entity' failed. I have selectively omitted various bindings, things and items, but have been unable to identify the trigger condition.
Only things I *think* make a difference are:
* Apparently I MUST delete the karaf cache in /var/lib/openhab2/cache each time. It does not work without (now I do this via Karaf property).
* when to load rules
I *believe* the 100% CPU core use only happens when there's rules.
When I move the rules files away before startup, startup seems to go fine. If I then move rules files back in, post-processing the rules takes much longer than when they're in at startup, BUT it succeeds.
The inverse is true for the time it takes to process things and items files: i.e. fast if done first, very slow when done after the rules. Likely TOO slow, and this makes us run into some timeout.

To move rules into the running OH (started without rules) currently seems to be the only startup procedure to get my OH running reliably.
Unfortunately, as I have a slow box and it's a production system, it's way too time consuming to try various options.
I'm running snapshot #1047 and the static jar zwave development binding right now, but it has been a problem for a long time, with the standard binding and across various snapshot versions and re-installations.

I've uploaded 3 startup session logs based on snapshot #1034 which I was on at that time, hoping that some merciful maintainer is capable of and willing to search and find a hint in there as to what's causing this lockup. On first start after #1034 install, I ended up with 100% CPU (no log there, sorry).
I moved away the rules and restarted, this is karaf1.txt. OH bundles didn't load, 
"Eclipse SmartHome Configuration Discovery" and "Eclipse SmartHome Bonjour/MDS Service Discovery Bundle" were stuck in Waiting state (rest was Active).
Another restart, this is karaf2.txt, same result. Then I removed the Karaf cache, this is karaf3.txt. This one worked in the end.
Note the zwave binding error messages are due to openhab-transport-serial bundle missing at startup.
I installed that one manually (btw, is there any option to make that happen automatically?).

There's two things to catch my eye:
1) many errors pointing at empty Maven XML files, I don't know if that's related to my problem (it's the first time I use karaf debug, so I don't know if this was there in earlier). I created separate [issue 544](https://github.com/openhab/openhab-distro/issues/544) for this.

2) OH bindings are loaded and even seem to work, but suddenly, for no apparent reason, bindings get uninstalled (!) starting with these lines:
2017-09-08 08:13:40.410 [INFO ] [internal.service.FeaturesServiceImpl] - Changes to perform:
2017-09-08 08:13:40.416 [INFO ] [internal.service.FeaturesServiceImpl] -   Region: root
2017-09-08 08:13:40.418 [INFO ] [internal.service.FeaturesServiceImpl] -     Bundles to uninstall:

Please please try to decode this puzzle for me.
Just ask when you want me to provide config or any other information.

[karaf1.txt](https://github.com/openhab/openhab-distro/files/1287175/karaf1.txt)
[karaf2.txt](https://github.com/openhab/openhab-distro/files/1287176/karaf2.txt)
[karaf3.txt](https://github.com/openhab/openhab-distro/files/1287174/karaf3.txt)

EDIT: I just discovered the "dev:dump-create" command in Karaf.
[Here](https://github.com/openhab/openhab-distro/files/1290391/2017-09-10_110654.zip)'s the output as of today.

**Comment by wborn** [comment]
According to the logging it looks like you are using KAR files with 2.2.0-SNAPSHOT for installation. So I think this issue is a duplicate of https://github.com/openhab/openhab-distro/issues/519.

### karaf1.txt

```
2017-09-08 08:12:34.078 [INFO ] [af.deployer.kar.KarArtifactInstaller] - Found a .kar file to deploy.
2017-09-08 08:12:34.081 [DEBUG] [mpl.info.InfoBundleTrackerCustomizer] - Ignore incorrect info null provided by bundle com.fasterxml.jackson.module.jackson-module-jaxb-annotations
2017-09-08 08:12:34.086 [INFO ] [af.deployer.kar.KarArtifactInstaller] - Installing KAR file /usr/share/openhab2/addons/openhab-addons-2.2.0-SNAPSHOT.kar
```

### karaf2.txt

```
2017-09-08 09:06:17.989 [INFO ] [af.deployer.kar.KarArtifactInstaller] - Found a .kar file to deploy.
2017-09-08 09:06:17.998 [INFO ] [af.deployer.kar.KarArtifactInstaller] - Installing KAR file /usr/share/openhab2/addons/openhab-addons-2.2.0-SNAPSHOT.kar
```

### karaf3.txt

```
2017-09-08 09:26:56.601 [INFO ] [af.deployer.kar.KarArtifactInstaller] - Found a .kar file to deploy.
2017-09-08 09:26:56.606 [INFO ] [af.deployer.kar.KarArtifactInstaller] - Installing KAR file /usr/share/openhab2/addons/openhab-addons-2.2.0-SNAPSHOT.kar
```

**Comment by wborn** [comment]
@mstormi does everything work correctly when using an online installation (not using KAR files)? If so, can you check if the other issues that you have opened and linked to this issue are also solved when not using KAR files?

**Comment by mstormi** [comment]
Thanks for checking into this.
No, unfortunately .kar ain't the reason. I had moved to online-repo-only installation some weeks ago, but the problem had persisted.
I don't think this is related to .kar because it only seems to happen whenever rules are loaded before items are (and I have been avoiding that by temporarily manually moving away the rules files), 
Also, I just restarted OH but failed to reproduce my issue. But then again, I'm not sure if everything's alright, and more important, this restart was the first time in a (felt) gazillion years that .items were loaded before .rules were.
As I believe ESH#1896 still ain't fixed (to enforce load order), so maybe I was just lucky this startup run.

Either way, given the area and number of code changes since I have opened this issue, I feel it does not make sense to keep hunting it on this old code+log basis.
May I ask you to keep this issue opened but just wait. I'll try to reproduce my issue and get new logs in case I'm successful in reproducing.

**Comment by wborn** [comment]
Could you reproduce this issue with new logs @mstormi? 

In the meantime I've created a [community post](https://community.openhab.org/t/openhab2-java-cpu-100/19205/12?u=wborn) about finding the CPU time per thread.

**Comment by mstormi** [comment]
No I couldn't reproduce/take logs.
I haven't seen this 100% CPU problem again for a long time.
Then again,  I'm now on #1248 and there have been substantial changes (.kar fixes, new Karaf, ...) since I opened this issue.
I'm still using some mechanics to move away .rules on startup and back, but it's merely to play things safe. I do have seen startup succeed without this.
Also, nowadays I get output such as 'refreshing xxx.rules' when a .items file is processed and the item is used in xxx rules, so I know there's progress.

I'll close this issue for now as not reproducible and will reopen or open a new one if needed. Will make use of your post on threading.
